### PR TITLE
mgr/dashboard: Review Section for the Create Cluster Workflow

### PR DESCRIPTION
**Pull request by avanthakkar**
Fixes: https://tracker.ceph.com/issues/50566
Signed-off-by: Avan Thakkar athakkar@redhat.com

Review page for create-cluster workflow:
Screenshot from 2021-07-16 15-26-55

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by nizamial09 — CHANGES_REQUESTED** [review]
I would create a new folder like `create-cluster-review` inside the `create-cluster` folder and put all the review components inside that.

**Comment by avanthakkar** [comment]
> I would create a new folder like `create-cluster-review` inside the `create-cluster` folder and put all the review components inside that.

Ok..but doesn't create-cluster folder should cover all the components for the workflow? As the create-cluster defines not just adding host but also other steps will get included later and the review section for the completion. I'm not sure though what suits better here. What you think @epuertat ?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.spec.ts
Comment (avanthakkar) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.spec.ts:35]: Forgot to change the description..will update it.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.html
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.html:18]: `Hosts by Label` or `Hosts per Label` ?

**Review by nizamial09 — COMMENTED** [review]
Instead of making the Host Details table small and aligned just to the right, how about making it bigger. I know that currently it has not useful information available but when we gather facts there will be. 

**With the current approach, in future this table will be congested.**
Screenshot from 2021-07-13 15-28-57


**I would suggest something like this.**
Screenshot from 2021-07-13 15-33-06

**Later on this can be like this**
Screenshot from 2021-07-13 15-31-30

But of course, this can be a follow up or iteration once this is merged. your call @avanthakkar

**Comment by avanthakkar** [comment]
> Instead of making the Host Details table small and aligned just to the right, how about making it bigger. I know that currently it has not useful information available but when we gather facts there will be.
> 
> **With the current approach, in future this table will be congested.**
> Screenshot from 2021-07-13 15-28-57
> 
> **I would suggest something like this.**
> Screenshot from 2021-07-13 15-33-06
> 
> **Later on this can be like this**
> Screenshot from 2021-07-13 15-31-30
> 
> But of course, this can be a follow up or iteration once this is merged. your call @avanthakkar

Yeah I was thinking the same, to expand the table once I start introducing the gather facts in the table and I was planning to introduce them in a separate PR , but yeah I can start working on it now and if possible introduce it in this PR itself.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.html
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.html:18]: To be honest i don't know...I thought maybe `Per` is better. But I can change it.

**Comment by nizamial09** [comment]
> but yeah I can start working on it now and if possible introduce it in this PR itself

Seperate PR is fine by me.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.html
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.html:18]: Hosts by Label I guess since we are filtering the Hosts by Label. and b of the by will be a small letter.

**Comment by nizamial09** [comment]
> I'm not sure though what suits better here. What you think @epuertat ?

@epuertat Can you let us know what do you think of this?

**Comment by epuertat** [comment]
> > I'm not sure though what suits better here. What you think @epuertat ?
> 
> @epuertat Can you let us know what do you think of this?

@nizamial09 @avanthakkar yeah I'm ok with that.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.html
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.html:29]: While checking at this layout, I'm wondering if this is going to be easy to brand with PF. The [PF wizard](https://www.patternfly.org/v4/components/wizard/html) relies on `<main>` and `<footer>` components together with specific styles for `...main`, `...main-body`, `...footer`, etc.

@nizamial09 did you check that the current wizard is going to work with the PF one?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.ts
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.ts:19]:
```suggestion
  labelOccurences = {};
```

**Review by epuertat — APPROVED** [review]
LGTM... But I'd suggest (not in this PR but in a clean-up one) that you both 'polish' a bit the UI (please check [the PF one as a reference](https://www.patternfly.org/v4/components/wizard/html)): 
* The numbers in the circles are slightly vertically misaligned.
* The current step (number should be highlighted).
* The padding on the left-side list of steps should be corrected too.
* I'd simplify tables by removing the top filtering tools (and would set a fixed number of max items, instead of the customizable 10).

**Comment by nizamial09** [comment]
> LGTM... But I'd suggest (not in this PR but in a clean-up one) that you both 'polish' a bit the UI (please check [the PF one as a reference](https://www.patternfly.org/v4/components/wizard/html)):
> 
> * The numbers in the circles are slightly vertically misaligned.
> * The current step (number should be highlighted).
> * The padding on the left-side list of steps should be corrected too.

Sure @epuertat There's a bit of cleanup needs to be done in the UI elements. I'll do that in a cleanup.

> * 'Review' title should be larger or more emphasized than the subtitles 'Cluster Resources', etc.

I think the Cluster Resources should be smaller. 'Cluster Resources' and all similar headings. That's how it is in the pf designs.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.html
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.html:29]: Since for pf mostly we are doing the scss stylings the wizard and its stepper will work fine. And this is a component that we call inside the wizard. So this will work like the rest of the component where the branding styles are already being applied. but I am guessing some work will needs to be done in downstream. @epuertat

**Comment by nizamial09** [comment]
@epuertat I did the UI improvments in here https://github.com/ceph/ceph/pull/42296

**Comment by aaSharma14** [comment]
jenkins test make check

**Review by aaSharma14 — APPROVED** [review]
Great Work @avanthakkar

**Review by Waadkh7 — APPROVED** [review]

**Review by nizamial09 — APPROVED** [review]
LGTM! Great job @avanthakkar

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.ts
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.ts:21]: Typo?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.ts
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.ts:16]:
```suggestion
  hostsDetails: object;
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.ts
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.ts:17]:
```suggestion
  hostsByLabel: object;
```

**Review by alfonsomthd — APPROVED** [review]

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.ts
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.ts:16]: It gives me error 
``` 
src/app/ceph/cluster/create-cluster/create-cluster-review.component.ts(84,25): error TS2339: Property 'data' does not exist on type 'object'.
[lint:tsc     ] src/app/ceph/cluster/create-cluster/create-cluster-review.component.ts(85,25): error TS2339: Property 'data' does not exist on type 'object'.
```
So I think I need to keep `any` type in the same way as [monitor component](https://github.com/ceph/ceph/blob/aeafb8dd4613d22e00ab9d9b2f05e0ca29cac6c5/src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/monitor/monitor.component.ts#L15)

Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.ts:16]: You have 2 options:
1) Create the appropriate data type for that var.
2) Use brackets instead of <variable>.<property>

Comment (epuertat) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.ts]: Good catch @alfonsomthd ! 2 comments from my end:
1. To avoid having to manually look for this, why not enabling [`no-any` tslint rule](https://palantir.github.io/tslint/rules/no-any/)?
2. I think `object` is better than `any` but not that much. @avanthakkar don't we have a `HostDetails` interface or type? Tslint `ban-types` allows [to reject certain types, like object](https://palantir.github.io/tslint/rules/ban-types/).

Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.ts]: I think for now declaring object type would work and using brackets instead of dots as @alfonsomthd sugeested.

Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.ts]: @epuertat I'm in for enforcing a rule, but not from TSLint - which is deprecated (we have to move to ESLint).
Also, enforcing this will require a dedicated PR that includes all `any` replacements.

Comment (epuertat) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.ts]: > I think for now declaring object type would work and using brackets instead of dots as @alfonsomthd sugeested.

@avanthakkar from a typing perspective, IMHO declaring something as `object` is just self-deception. You are not enforcing any real data structure, so there's no big gain between `any` and `object`.

@alfonsomthd the fact that TSLint is deprecated doesn't mean that the rules it enforces are deprecated too. I'd rather put this rule in place right now rather than deferring it because we have to update the whole tooling and that's not trivial... Personally I don't think we should spend time reviewing things that any tool (deprecated or not) can detect for us.

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by avanthakkar** [comment]
jenkins render docs

**Pull request merged**
